app/models/champs/epci_champ.rb

```diff
@@ -95,8 +95,14 @@ def external_id_in_departement_epci_codes
   end
 
   def value_in_departement_epci_names
-    return if value.in?(APIGeoService.epcis(code_departement).pluck(:name))
-
-    errors.add(:value, :not_in_departement_epci_names)
+    # byebug
+    if value.in?(APIGeoService.epcis(code_departement).pluck(:name))
+      return
+    elsif external_id.in?(APIGeoService.epcis(code_departement).pluck(:code))
+      self.value = (external_id)
+      save
+    else
+      errors.add(:value, :not_in_departement_epci_names)
+    end
```

Comment:
proposition de code qui me paraît plus lisible (moins de branche, on pourrait d'ailleurs activer le cop pour aller dans ce sens)

```suggestion
    return if value.in?(APIGeoService.epcis(code_departement).pluck(:name))

    if external_id.in?(APIGeoService.epcis(code_departement).pluck(:code))
      self.value = external_id
      save
    else
      errors.add(:value, :not_in_departement_epci_names)
    end
```


sinon dans le fond même si ça débloque à court terme les instructeurs, je me demande si ça va pas poser des problèmes de faire un save dans un validate. Ce serait peut-être plus safe et logique de le faire dans un `before_validation`, voir un `before_save` . Et probablement qu'il n'y aurait pas besoin de faire de save, seulement l'assignation, car ce callback est systématiquement appelé dans le contexte d'un save (et je pense qu'on se fiche d'assigner ce genre de valeur hors contexte de save). En quelque sorte, il s'agit de normalisation de données, comme quand on fait un `strip` ou autre ailleurs avec `normalize`

Reply:
refactoré pr un truc plus propre : https://github.com/demarches-simplifiees/demarches-simplifiees.fr/pull/9888/files tx!